docs/design.md

```diff
@@ -0,0 +1,115 @@
+  - `Pod` sweeper: A periodical task runner to sweep Pods which exceeds the deletion
+    time limit.
+- Slack agent: Server for notifying user whether jobs are failed or not and
+  accepting extending requests from outside of Kubernetes.
+
+Diagram: T.B.A
+
+Architecture
+------------
+
+### How self-hosted runners are created and runs jobs
+
+1. The `RunnerPool` reconciler watches `RunnerPool` creation events and creates
+  a `Deployment` which own runner `Pod`s. The controller adds some labels to
+  `Pod`s to tell other components like webhook that the `Pod`s are managed by
+  the controller. After this steps, Other components recognize which `Pod`s
+  they should handle with those labels.
+1. The mutating webhook creates registration tokens via GitHub API and injects
+  them to the `Pod`s' `env` fields.
+1. Each runner `Pod` runs the following commands.
+  1. Register itself as a self-hosted runner with the injected token.
+  1. Initialize runner environment by doing the user-defnied process.
+  1. Start a long polling process and wait for GitHub Actions to assign a job.
+  1. Run an assigned job.
+  1. Call the Slack agent to notify users.
+    - GitHub API does not seem to provide a way to know which runner ran a
+      succeded or failed job.
+    - This repository provides a simple `job-failed` command, and asks users to
+      execute this command when the job is failed. The `if: failure()` syntax
+      allows users to run a step only when one of previous steps exit with non-zero
+      code.
+  1. Annotate the `Pod` manifest for itself with a timestamp when to delete this
+    `Pod`.
+    - If the job is succeded or canceled, the `Pod` annotates itself with the
+      current time.
+    - If the job is failed, the `Pod` annotate itself with the future time, for
+      example 20 min later.
+1. The Slack agent notifies the result of the job on a Slack channel.
+1. Users can extend the failed runner if they want to by clicking a button on Slack.
+1. The Slack agent is running a WebSocket process to watch extending messages.
+  If it receives a message, it annotate the `Pod` manifest with the designated time.
+1. `Pod` sweeper periodically checks if there are `Pod`s annotated with the old
+  timestamp, and if any, it deletes `Pod`s.
+
+### Runner's state management
+
+A runner `Pod` has the following state as a GitHub Actions job runner.
+
+- registerred: `Pod` registerred itself on GitHub Actions.
+- initialized: `Pod` finished the initialization.
+- listening: `Pod` starts listening with Long Polling.
+- assigned: `Pod` is assigned a job and starts running it.
+- debug: The job has finished with failure and Users can enter `Pod` to debug.
+
+Runner `Pod`s might not need to manage all of these state, but some of thes states
+are useful to check the system is properly running or not.
+The state "debug" is a must to be watched to keep runner `Pod`s in stock.
+Users can leave many failed runner `Pod`s for investigation, but the number of
```

Comment:
If all the replicas are occupied for the user's debug, does that mean that the self-hosted runner will not run workflow anymore?

Reply:
Yes, but autoscaling might help us avoid this situation. It's future work.